### PR TITLE
Loader in the flow board

### PR DESCRIPTION
**Pull request by matrix-amiel**
Hi.

We have found that flow board loaded very slowly in the large amount of appointments -
Discussion about this problem - https://community.open-emr.org/t/flow-board-is-very-slow/9781 

I made a small change in the logic of the flow board code and I load the data using ajax also in the first loading. With this change It's possible to add loader while the loading time (only for first loading or manual loading not for loading by timer).

while developing I found that the summaries of the appointments in the title of table are not refreshed in the ajax process (by filter or timer) and I fixed it.     

This is necessary fix because without the loader the user sees blank page until all the appointments are loaded and he don't know why it not work..

Thanks
Amiel

### CODE REVIEW

**Comment by mdsupport** [comment]
Just curious, using ajax on open makes the script more usable (saves one click) but doesn't address performance issue, correct?

I suspect getting this little baby to display 300 appts every 30 seconds on several devices will play havoc at the server.  And even with sub-second response for 100s of appointments, a human user will struggle to digest that dynamic data.

**Comment by matrix-amiel** [comment]
Right, it's only user experience fix. 
I agree there is serious performance problem in the big amount of appointments, and therefore we've canceled the automatic refresh in our clinics..

What do you suggest? this is very problematic that clinic can work only with several dozen appointments.. the flow board is one of the main screens in the system..

**Comment by mdsupport** [comment]
Couldn't agree with you more.  We have this as 'app'/standalone window on every user's device so everyone is aware of the queue without interrupting others.

Performance problems for this script stem from the script logic and emr design issues.  
- Primary purpose of this script is to track today's encounters and its activities - but it uses appointments as the source.
- fn that builds list of appointments is expensive and uses full scan (so limits don't speed up).

So one of the option to consider would:
1. Force non-admin users to define their universe better by saving their user setting so that their list is limited to about 30 or so tracks.  E.g. front desk should select facility they are handling, physicians should choose `their` patients at the facility they are at `today`, vitals technician should select patients in waiting room etc..
2. Move current strange ajax update out of current script and limit it to scanning for tracker element records since last request filtered by user settings.
3. ajax update should provide only id and status fields - status, room, user and time of last update.  The post response then updates just the changes.  This implies even if you have 300 items initialized on screen, you would receive only 1-2 updates every minute.
4. Refresh entire screen on a lazy cycle or if an id is not found in current display. This will fetch new appointments/encounters.
5. A separate timer should keep updating the clocks locally on the device.

We have rolled out some of these and waiting for user feedback.

**Comment by matrix-amiel** [comment]
Hi @mdsupport  . 
I agree with your solutions and also think that we need to retrieve only the new changes.
I have thought also on cache solution to improve performance.  

However this it requires serious plan and development and we can not do it in soon, Here I only try to resolve the immediate problem in our clinic, The work style of our clinic requires hundreds of appointments for single user (therefore the first solution not match for me..), I would like to avoid the blank page until all the appointments are loaded and to give feedback to user that screen is working.

I think my change is minor and not changes the exist logic and hopefully it can merge until will do refactor on this screen.

**Comment by mdsupport** [comment]
Hi @matrix-amiel, by no means I was suggesting turning this PR into a major project.  As I noted in original comment, I was just trying to validate my impression vs the issue referenced.
If the default parameters take care of even 20% of anticipated users, forcing everyone to confirm report parameters never makes sense.

File: interface/patient_tracker/patient_tracker.php [hunk 1]
```diff
@@ -434,12 +420,39 @@ class="btn btn-primary"
...
+    foreach ($appointments as $apt) {
+        $chk_prov[$apt['uprovider_id']] = $apt['ulname'] . ', ' . $apt['ufname'] . ' ' . $apt['umname'];
+    }
+
+                ?>
+                <div class="col-sm-12 text-center" style='margin:5px;'>
                 <span class="hidden-xs" id="status_summary">
                     <?php
-                    $statuses_output = "<span style='margin:0 10px;'><em>" . xlt('Total patients') . ':</em> <span class="badge">' . text($appointments_status['count_all']) . "</span></span>";
+                    $statuses_output = "<span style='margin:0 10px;'><em>" . xlt('v  patients') . ':</em> <span class="badge">' . text($appointments_status['count_all']) . "</span></span>";
```
Comment (bradymiller) [inline interface/patient_tracker/patient_tracker.php]: What is 'v patients' ??

Comment (matrix-amiel) [reply interface/patient_tracker/patient_tracker.php]: ooops.. 
fixed.

**Comment by bradymiller** [comment]
hi @matrix-amiel ,
This is testing well and will bring this into the codebase. Thanks for the improvement! -brady

**Pull request merged**